### ISSUE TITLE
yoloworldv2 WorldTrainerFromScratch 复现时loss降很慢

### ISSUE THREAD

**Issue by ccl-private** [ultralytics/ultralytics#21118]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

```python

from ultralytics import YOLOWorld
from ultralytics.models.yolo.world.train_world import WorldTrainerFromScratch

data = {
    "train": {
        "yolo_data": ["Objects365.yaml"],
        "grounding_data": [
            {
                "img_path": "./datasets/flickr30k/images",
                "json_file": "./datasets/flickr30k/final_flickr_separateGT_train.json",
            },
            {
                "img_path": "./datasets/GQA/images",
                "json_file": "./datasets/GQA/final_mixed_train_no_coco.json",
            },
        ],
    },
    "val": {"yolo_data": ["lvis.yaml"]},
}

model = YOLOWorld("yolov8s-worldv2.yaml")
model.train(data=data, batch=128, epochs=100, device=[0, 1], trainer=WorldTrainerFromScratch, name="test_yoloworldv2")
```

我用这个从头训练,目前指标看上去还很低,照这个趋势到100轮很难有很大的提升.
![Screenshot From 2025-06-20 11-28-05](https://github.com/user-attachments/assets/3d2957f6-f9b9-4f26-b9f1-059c8674236e)
官方的模型是这些参数训练出来的吗?如果不是能不能给一些你们训练时的详细参数,或者训练日志.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ccl-private, thank you for reaching out and sharing your experience with training yoloworldv2 from scratch 🚀!

We recommend reviewing the [Docs](https://docs.ultralytics.com/) for best practices, where you’ll also find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples. For custom training questions like this, please provide as much detail as possible, including dataset image samples and full training logs. Also, double-check our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to ensure optimal setup.

If you believe this is a 🐛 Bug Report, please add a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug further.

Join the Ultralytics community for real-time help or discussion:
- [Discord](https://discord.com/invite/ultralytics) 🎧
- [Discourse](https://community.ultralytics.com/)
- [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Please ensure you are using the latest `ultralytics` package, along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml), in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

You can run YOLO in any of these up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

🔔 This is an automated response. An Ultralytics engineer will review your issue and provide further assistance soon!

**Comment by glenn-jocher** [comment]
Looking at your training setup, it matches our [official YOLO-World training documentation](https://docs.ultralytics.com/models/yolo-world/) quite closely. Slow loss reduction in the early stages is normal for training from scratch, especially with mixed datasets including both detection and grounding data.

The training parameters you're using (batch=128, epochs=100) are consistent with our recommended setup. However, training from scratch typically requires patience as the model needs to learn both vision and language representations simultaneously. Consider monitoring for longer - significant improvements often occur after epoch 20-30.

You might also want to experiment with the learning rate (try `lr0=1e-4` or `2e-4`), ensure your datasets are properly formatted, and verify that both Objects365 and grounding datasets are loading correctly. The WorldTrainerFromScratch is specifically designed for this mixed training approach, so your setup looks correct.

**Comment by ccl-private** [comment]
![Image](https://github.com/user-attachments/assets/25b7ab80-57d2-482c-a8cd-8ba5c208fd13)
我按教程yoloworldv2 WorldTrainerFromScratch的数据集和参数设置训练好了100轮的模型,但P才0.45,R才0.13

**Comment by glenn-jocher** [comment]
Those results (P=0.45, R=0.13) are indeed quite low after 100 epochs. This suggests there may be an issue with your dataset preparation or training configuration rather than just needing more time.

Please verify that your datasets are loading correctly by checking the training logs for proper sample counts from both Objects365 and grounding datasets. Also ensure your annotation files match the expected format exactly - incorrect label formatting is a common cause of poor training performance with WorldTrainerFromScratch.

Try reducing your batch size to 64 or 32 and adding `patience=30` to allow for more training variations. You can also validate your setup by first training on a smaller subset to confirm the pipeline works before scaling to the full datasets.

**Comment by ccl-private** [comment]
下面是3对val的label和pred图,我看lable是对的,是不是可以证明图片准备没什么问题,我的数据源完全是照着官方教程准备的
label1:

![Image](https://github.com/user-attachments/assets/1afc9d0f-2d9c-4cd3-80bf-3bc068aa3b68)
pred1:

![Image](https://github.com/user-attachments/assets/153c5c30-10a7-4c91-83d9-0e55b7682c5b)
label2:

![Image](https://github.com/user-attachments/assets/b0626363-d33a-4ea6-a903-054386d9db6b)
pred2:

![Image](https://github.com/user-attachments/assets/9b75b2a2-63bd-4c01-890e-0eac14b11a10)
label3:

![Image](https://github.com/user-attachments/assets/d0ad3597-366e-4e65-8292-5f9612065d60)
pred3:

![Image](https://github.com/user-attachments/assets/a6d1dfc7-d74b-4375-9239-4ed29391109b)

**Comment by glenn-jocher** [comment]
Your validation labels look correct, which confirms your data preparation is fine. The stark difference between labels and predictions indicates the training process itself needs adjustment rather than data issues.

For YOLO-World training from scratch, try these specific modifications: reduce your learning rate significantly (use `lr0=5e-5` or `lr0=1e-4`), add warmup with `warmup_epochs=5`, and consider using a smaller batch size like `batch=32` for more stable gradients. The mixed detection/grounding training in WorldTrainerFromScratch is particularly sensitive to hyperparameters.

Also check your training logs to verify both detection and grounding losses are decreasing - if one loss dominates, the model may not learn properly. You might want to train for longer (200+ epochs) as YOLO-World from scratch typically requires more iterations than fine-tuning pretrained models.
